merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/TypePattern.java

```diff
@@ -15,71 +16,72 @@
-      case ARRAY: return new ArrayPattern(TypePattern.from(((ArrayType) mirror).getComponentType()));
-      case TYPEVAR: return new TypeVariablePattern(mirror.toString());
-      case DECLARED: {
+      case BOOLEAN -> result = new PrimitivePattern(Primitive.BOOLEAN);
+      case BYTE -> result = new PrimitivePattern(Primitive.BYTE);
+      case SHORT -> result = new PrimitivePattern(Primitive.SHORT);
+      case INT -> result = new PrimitivePattern(Primitive.INT);
+      case LONG -> result = new PrimitivePattern(Primitive.LONG);
+      case CHAR -> result = new PrimitivePattern(Primitive.CHAR);
+      case FLOAT -> result = new PrimitivePattern(Primitive.FLOAT);
+      case DOUBLE -> result = new PrimitivePattern(Primitive.DOUBLE);
+      case ARRAY -> result = new ArrayPattern(TypePattern.from(((ArrayType) mirror).getComponentType()));
+      case TYPEVAR -> {
+        var typeVarName = mirror.toString();
+        if (typeVarName.contains(" ")) {
+          typeVarName = typeVarName.substring(typeVarName.lastIndexOf(' ') + 1);
+        }
+        result = new TypeVariablePattern(typeVarName);
+      }
+      case DECLARED -> {
         // DeclaredType element can be cast as TypeElement because it's a Type
         final var className = ClassName.get((TypeElement) ((DeclaredType) mirror).asElement());
         final var typeArguments = ((DeclaredType) mirror).getTypeArguments();
         final var argumentPatterns = new ArrayList<TypePattern>(typeArguments.size());
         for (final var typeArgument : typeArguments) {
           argumentPatterns.add(TypePattern.from(typeArgument));
         }
-        return new ClassPattern(className, argumentPatterns);
+        result = new ClassPattern(className, argumentPatterns);
       }
 
-      default:
-        throw new Error("Cannot construct a pattern for type " + mirror);
+      default -> throw new Error("Cannot construct a pattern for type " + mirror);
     }
+
+    final var annotations = mirror.getAnnotationMirrors();
+    for (int i = annotations.size() - 1; i >= 0; i--) {
+      final var annotation = annotations.get(i);
+      final var className = ClassName.get((TypeElement) annotation.getAnnotationType().asElement());
+      result = new AnnotationPattern(className, Optional.of(annotation), result);
+    }
+    return result;
   }
 
-  public static TypePattern from(final VariableElement element) {
+  static TypePattern from(final VariableElement element) {
     return from(element.asType());
   }
 
-  public abstract TypePattern substitute(Map<String, TypePattern> substitution);
-
-  public abstract Map<String, TypePattern> match(TypePattern other) throws UnificationException;
+  TypePattern substitute(Map<String, TypePattern> substitution);
 
-  public abstract boolean isGround();
+  Map<String, TypePattern> match(TypePattern other) throws UnificationException;
 
-  public abstract boolean isSyntacticallyEqualTo(TypePattern other);
+  boolean isGround();
```

Comment:
As we discussed verbally, these methods could really use some javadoc. Let me know what you think:

```java
  /**
   * Given a TypePattern, produce a new TypePattern where all type variables whose names are
   * in the given map have been replaced with the corresponding type patterns.
   * @param substitution A map from type variable name to the type pattern with which to replace it
   * @return a TypePattern with the same structure, but with any matching type variables replaced
   * E.g. ClassPattern(List, TypeVariablePattern(T)).substitute(Map.of("T", ClassPattern(Double))) will return
   *   ClassPattern(List, Double)
   */
  TypePattern substitute(Map<String, TypePattern> substitution);

  /**
   * Attempt to match this TypePattern against the given other TypePattern. Throw a UnificationException if they cannot be matched
   * @param other the ground TypePattern to match against this TypePattern
   * @return an assignment of type variables to TypePatterns, if match is successful
   * @throws UnificationException if the two patterns do not match
   */
  Map<String, TypePattern> match(TypePattern other) throws UnificationException;

  /**
   * Whether this TypePattern contains any type variables. Ground is true if there are no type variables present.
   * @return false if there is at least one type variable in this TypePattern. true if there are none.
   */
  boolean isGround();

  /**
   * Whether this TypePattern has the same structure, same types, and same type variable names as the other pattern
   * @param other the other TypePattern to compare with
   * @return true if the two TypePatterns are identical, false otherwise.
   */
  boolean isSyntacticallyEqualTo(TypePattern other);

  /**
   * Represent this TypePattern as a javapoet TypeName
   * @return a javapoet TypeName corresponding to this TypePattern
   *
   * E.g. ClassPattern(List, Double).erasure --> ClassName(List, Double)
   */
  TypeName render();

  /**
   * Represent this TypePattern as a javapoet TypeName without generics
   * @return a javapoet TypeName corresponding to this TypePattern without generics
   *
   * E.g. ClassPattern(List, Double).erasure --> ClassName(List)
   */
  TypeName erasure();

  /**
   * If this TypePattern represents a primitive, return a TypePattern that represents the corresponding Object type.
   * @return a TypePattern that is not a PrimitiveTypePattern
   *
   * E.g. PrimitivePattern(boolean) --> ClassPattern(Boolean)
   *
   * There is no need to recurse: if the top level is not a PrimitiveTypePattern,
   * the entire type pattern is already considered "boxed"
   */
  TypePattern box();
```

Reply:
This is perfect, thanks! One note: the example in `render`'s JDoc says `.erasure` instead of `.render`